compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -684,9 +682,171 @@ class ExpressionCodegen(
 
     override fun visitWhen(expression: IrWhen, data: BlockInfo): StackValue {
         expression.markLineNumber(startOffset = true)
+        val switchGen = SwitchGenerator(expression, data, intrinsics)
+        if (switchGen.match())
+            return switchGen.gen()
         return genIfWithBranches(expression.branches[0], data, expression.type.toKotlinType(), expression.branches.drop(1))
     }
 
+    // TODO: Char, Byte, Short, String, enum, eliminate the temporary variable
+    inner class SwitchGenerator(val expression: IrWhen, val data: BlockInfo, val intrinsics: IrIntrinsicMethods) {
```

Comment:
Can we extract this to a separate file? (Would be nice to avoid making this ExpressionCodegen as enormous as the one in the old backend :) )

Reply:
Done. Please take a look again :)